### PR TITLE
Shallow copy parent and children in DataTree constructor

### PR DESCRIPTION
**Pull request by TomNicholas**
This seems to fix the examples in #9196, but causes other tests to fail, which I think just reveals that we have existing tests that assume this in-place modification is desired. I will change those tests to match the new intended behaviour.

- [x] Implements suggestion (1) from https://github.com/pydata/xarray/issues/9196#issue-2382534068
- [x] Tests added
- [ ] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
- [ ] New functions/methods are listed in `api.rst`

### CODE REVIEW

File: xarray/core/datatree.py [hunk 1]
```diff
@@ -463,8 +463,10 @@ def __init__(
 
         super().__init__(name=name)
         self._set_node_data(_coerce_to_dataset(data))
-        self.parent = parent
-        self.children = children
+
+        # shallow copy to avoid modifying arguments in-place (see GH issue #9196)
+        self.parent = parent.copy() if parent is not None else None
```
Comment (shoyer) [inline xarray/core/datatree.py:468]: Can we remove `parent` from the constructor instead?

It's a little weird to only be able to set the _direct_ parent (and corresponding sibling), but not more distant relatives.

Comment (TomNicholas) [reply xarray/core/datatree.py:468]: Yeah I was going to do that in a follow-up PR, but I can do it in this one instead.

**Comment by TomNicholas** [comment]
The main slowdown here is just how some of the tests are written - a lot of them rely upon in-place modification to work. e.g. 

https://github.com/pydata/xarray/blob/c7cb9f7eadb750618ea11121c2741b331fdf966c/xarray/tests/test_datatree.py#L77

This change makes the constructor a somewhat unhelpful way to create trees from scratch. I could rewrite these tests either

1) The ugly way I have started in https://github.com/pydata/xarray/pull/9297/commits/5db25bd19246d282e6b3a348bacded0f8dd40369, which doesn't refer to copied objects,

2) Using `.from_dict`, which is much clearer (and the recommended user API), but relies on more complex code paths internally so doesn't isolate the behaviour of the constructor / relationship properties so much.

**Comment by flamingbear** [comment]
> 1. The ugly way I have started in [5db25bd](https://github.com/pydata/xarray/commit/5db25bd19246d282e6b3a348bacded0f8dd40369), which doesn't refer to copied objects,
 
When you say it's ugly, it's what I was thinking.  It seems closer to what you're testing. 
This looks like a not super useful test. 

```
john = DataTree.from_dict(
    {
        "/": DataTree(name="John"),
        "/Mary": DataTree(name="Mary"),
        "/Mary/Sue": DataTree(name="Sue"),
    }
)
assert john["/Mary/Sue"].path == "/Mary/Sue"
 
```

**Comment by shoyer** [comment]
I think using `from_dict` for tests is fine, as long as you thoroughly test the constructor as well.

You could also avoid all the intermediate variables by writing something like:
```python
john = DataTree(children={"Mary": DataTree(children={"Sue": DataTree()})})
```

Incidentally, it would sure be nice if we could eventually support automatic unpacking of nested dictionaries in the `DataTree` constructor, e.g., something like:
```python
john = DataTree({"Mary": {"Sue": {}}})
```
Here the nature of the node (child DataTree vs DataArray) could be inferred from the argument type.

**Comment by shoyer** [comment]
> ```python
> john = DataTree.from_dict(
>     {
>         "/": DataTree(name="John"),
>         "/Mary": DataTree(name="Mary"),
>         "/Mary/Sue": DataTree(name="Sue"),
>     }
> )
> assert john["/Mary/Sue"].path == "/Mary/Sue"
>  ``

Incidentally, I would try to avoid redundant `name` arguments, given how `name` is (or should be) overriden by the assigned key in the parent object. So I would wrtie this as:
```python
john = DataTree.from_dict(
    {
        "/": DataTree(),
        "/Mary": DataTree(),
        "/Mary/Sue": DataTree(),
    }
)
```
or even just
```python
john = DataTree.from_dict(
    {
        "/Mary/Sue": DataTree(),
    }
)
```

**Comment by TomNicholas** [comment]
I've now re-written the tests to use approach (2), i.e. to just use `DataTree.from_dict` everywhere that a nested tree is required.

**Comment by flamingbear** [comment]
@TomNicholas I didn't want to push up unfinished work to this PR again. But https://github.com/pydata/xarray/commit/7c530f5a3c09c7331b30db6bb7e5a790c754e192?diff=split&w=0
is my stab at removing the parent parameter from the DataTree Constructor.  I got all but three tests to pass but it seems like the three might not make sense anymore (or I just don't fully understand them)

```
FAILED xarray/tests/test_datatree.py::TestFamilyTree::test_create_two_children - TypeError: DataTree.__init__() got an unexpected keyword argument 'parent'
FAILED xarray/tests/test_datatree.py::TestFamilyTree::test_dont_modify_parent_inplace - TypeError: DataTree.__init__() got an unexpected keyword argument 'parent'
FAILED xarray/tests/test_datatree.py::TestFamilyTree::test_setparent_unnamed_child_node_fails - TypeError: DataTree.__init__() got an unexpected keyword argument 'parent'
```

**Comment by TomNicholas** [comment]
I think these are all resolved by either:
1. Changing outdated tests
2. Also making `.parent` read-only.

**Comment by TomNicholas** [comment]
This PR now does a few related things:
1) Shallow copy `children` in the `DataTree` constructor (https://github.com/pydata/xarray/pull/9297/commits/6c56b12d4d62f8549824cd698fcd5991c5205eaa),
2) Removes the `parent` argument from the `DataTree` constructor (https://github.com/pydata/xarray/pull/9297/commits/7c530f5a3c09c7331b30db6bb7e5a790c754e192)
3) Makes the `.parent` attribute of `DataTree` read-only (https://github.com/pydata/xarray/pull/9297/commits/897b589239311e590a6d12fb287120c904c7f7a2)
4) Changes any tests that these effect by making them pass or deleting them.

**Comment by TomNicholas** [comment]
I believe this is ready to merge. It might imply more changes to @flamingbear 's docs PR though.

**Review by shoyer — APPROVED** [review]

**Comment by shoyer** [comment]
Looks great, thanks Tom!

**Pull request merged**